### ISSUE TITLE
NullReferenceException at Microsoft.AspNetCore.Http.Internal.DefaultHttpRequest.get_Query()

### ISSUE THREAD

**Issue by clement911** [aspnet/Home#2806]
I just got a weird null reference with the following call stack:

```
System.NullReferenceException: Object reference not set to an instance of an object.
at Microsoft.AspNetCore.Http.Internal.DefaultHttpRequest.get_Query()
...
```

The calling code looks like this

```
if (_httpAccessor.HttpContext?.Request.Query.ContainsKey(SharedConstants.HEADER_SHOP_ID) == true)
{
...
}
```

The app is running on .Net 4.6.1

  ```
...
<package id="Microsoft.AspNetCore.Http" version="2.0.0" targetFramework="net461" />
<package id="Microsoft.AspNetCore.Http.Abstractions" version="2.0.0" targetFramework="net461" />
<package id="Microsoft.AspNetCore.Http.Extensions" version="2.0.0" targetFramework="net461" />
<package id="Microsoft.AspNetCore.Http.Features" version="2.0.0" targetFramework="net461" />
...
```

**Comment by Tratcher** [comment]
There are some known issues where the internal HttpContext structure does not react well to multi-threaded access, especially to properties that lazily initialize like Query. It looks like you're using IHttpContextAccessor? Is there another thread that may be accessing this context concurrently?

**Comment by clement911** [comment]
I don't think there would be another thread accessing the same http request, but multiple threads may be using the accessor to access their own http request. Considering I saw this error only once, a threading  is certainly possible.
Is a fix on the roadmap?

**Comment by Tratcher** [comment]
No, the types are fundamentally not designed for multi-threaded use.

**Comment by clement911** [comment]
When you say _types_ are you referring to HttpContextAccessor or HttpRequest?

It seems reasonable for HttpRequest to not be thread safe but not for HttpContextAccesso (unless I really didn't understand its purpose).

If HttpContextAccessor was not thread safe it would be like saying that the old System.Web.HttpContext.Current was not thread safe. You know what I mean?

**Comment by Tratcher** [comment]
Yes, HttpContext. That said, we've often seen HttpContextAccessor used to accidentally access a request concurrently with another thread, such as inside loggers that aren't directly aware of the threading. Where are you using it?

**Comment by clement911** [comment]
Yes I am using it in a custom logger too, to get the user id from the request.
But I don't believe there is any code path that can lead to multiple threads logging a message **on the same request**. That is, for a given request, the logger.Log* calls should be serial.
Can I get in trouble anyway?

**Comment by Tratcher** [comment]
Some thread transitions are not obvious, such as HttpClient spinning up a background thread to process outgoing requests.

**Comment by Tratcher** [comment]
Are you starting any async operations that you don't immediately await? E.g. sending multiple HttpClient requests before awaiting the results?

**Comment by clement911** [comment]
It's hard for me to confirm because logging calls are scattered everywhere in the code base, but I'm certain the httpaccessor is only used from the logger.
I suppose I could lock the `_httpAccessor.HttpContext` in the logger before accessing its properties.

**Comment by Tratcher** [comment]
That will only help if the conflict is worth other logging, but it's worth a try. The other option is to pre-touch lazy properties like this at a known safe place such as in a middleware.

**Comment by clement911** [comment]
Got ya. Feel free to close this as you see fit.

**Comment by antonGritsenko** [comment]
Sorry for retake this topic, but don't want to open a new one. 
@Tratcher, you've mentioned that middleware is kind of safe place, but what about ActionFilters? 
We have a pretty easy one like this:
```
public class PartialViewFilter : IActionFilter
    {
        public void OnActionExecuted(ActionExecutedContext context)
        {
            
        }

        public void OnActionExecuting(ActionExecutingContext context)
        {
            context.HttpContext.Items["_partial"] = context.HttpContext.Request.Query.Keys.Contains("partial");
        }
    }
```
and catches NullReferenceException sometimes (not very often, less than 5% requests aprox).
Unfortunately, we still using old asp.net core (1.1), but more interesting that problem start to be occurs only when we moved from 1.0 to 1.1 (no other changes at all). 
Exception:

> System.NullReferenceException: Object reference not set to an instance of an object.
> at Microsoft.AspNetCore.Http.Internal.DefaultHttpRequest.get_Query()
> at ****.PartialViewFilter.OnActionExecuting
> …

We have a lot of places where HttpContext (custom authentication filters) and HttpContextAccessor used (custom logging mainly, but also a lot of accesses to Session). So the second question: is any recommendation on how to use HttpContext\Accessor safely?

Thank you

**Comment by Tratcher** [comment]
Comments on closed issues are not tracked, please open a new issue with the details for your scenario.

**Comment by davidfowl** [comment]
@clement911 @antonGritsenko  if you can still reproduce the issue, see if this guide helps capture a process dump https://github.com/microsoft/ApplicationInsights-dotnet/issues/1524#issuecomment-1146848272
